src/app/api/book/analyze/route.ts

```diff
@@ -0,0 +1,88 @@
+import { NextRequest, NextResponse } from "next/server";
+import { OpenAIEmbeddings, ChatOpenAI } from "@langchain/openai";
+import { RecursiveCharacterTextSplitter } from "@langchain/textsplitters";
+import { MemoryVectorStore } from "langchain/vectorstores/memory";
+import {
+  RunnablePassthrough,
+  RunnableSequence,
+} from "@langchain/core/runnables";
+import { StringOutputParser } from "@langchain/core/output_parsers";
+import { ChatPromptTemplate } from "@langchain/core/prompts";
+import type { Document } from "@langchain/core/documents";
+
+const formatDocumentsAsString = (documents: Document[]) => {
+  return documents.map((document) => document.pageContent).join("\n\n");
+};
+
+export async function POST(request: NextRequest) {
+  try {
+    const { fileContent } = await request.json();
+
+    if (!fileContent) {
+      return NextResponse.json(
+        { error: "File content is required" },
+        { status: 400 }
+      );
+    }
+
+    const model = new ChatOpenAI({
+      modelName: "gpt-4",
+    });
+
+    const textSplitter = new RecursiveCharacterTextSplitter({ chunkSize: 1000 });
+    const docs = await textSplitter.createDocuments([fileContent]);
+
+    // Create a vector store from the documents
+    const vectorStore = await MemoryVectorStore.fromDocuments(
+      docs,
+      new OpenAIEmbeddings()
+    );
+
+    // Initialize a retriever wrapper around the vector store
+    const vectorStoreRetriever = vectorStore.asRetriever();
+
+    // Create a system & human prompt for the chat model
+    const SYSTEM_TEMPLATE = `You are a literary analyst. Use the following pieces of context to answer the question at the end.
+    If you don't know the answer, just say that you don't know, don't try to make up an answer.
+    ----------------
+    {context}`;
+
+    const prompt = ChatPromptTemplate.fromMessages([
+      ["system", SYSTEM_TEMPLATE],
+      ["human", "{question}"],
+    ]);
+
+    const chain = RunnableSequence.from([
+      {
+        context: vectorStoreRetriever.pipe(formatDocumentsAsString),
+        question: new RunnablePassthrough(),
+      },
+      prompt,
+      model,
+      new StringOutputParser(),
+    ]);
+
+    const questions = {
+      keyCharacters: "Identify the key characters in the book and provide a list of their names.",
```

Comment:
**Suggesiton: using tools / function calling might give you an easier output structure to work w/ **

Can talk about more in person